### ISSUE TITLE
Invalid SQL when using distinct, group, limit and count

### ISSUE THREAD

**Issue by pascalbetz** [rails/rails#30886]
I believe the commit https://github.com/rails/rails/commit/0e8d4edd56f44b27f33bd214aa03af60cc5d39ae has introduced a regression.

Using `distinct`, `group`, `limit` and `count` (without an explicit `select`) causes an invalid SQL statement being generated.

### Steps to reproduce

Run following in a Rails console with 5.1.4 (error) and 5.1.3 (ok):

`SomeModel.group(:id).order('1 DESC').limit(30).distinct.count`

(in the samples below, my model was `Device`)

### Expected behavior

It should create a valid SQL statement and execute it. Something like:

`SELECT  COUNT(DISTINCT "devices"."id") AS count_id, "devices"."id" AS devices_id FROM "devices" GROUP BY "devices"."id" ORDER BY 1 DESC LIMIT $1`

### Actual behavior

Creates invalid SQL:

`SELECT  COUNT(DISTINCT *) AS count_all, "devices"."id" AS devices_id FROM "devices" GROUP BY "devices"."id" ORDER BY 1 DESC LIMIT $1`

The `SELECT  COUNT(DISTINCT *) ` fragment is invalid.

### System configuration

Postgres (9.6)

**Rails version**:
Fails with: Rails 5.1.4
Works with: Rails 5.1.3, 5.1.2

**Comment by pascalbetz** [comment]
Digging a bit further:
The condition that has `column_name` take the value of `primary_key` has changed, leading it to keep the value  `:all` which in turn leads to it being replaced by `*` later on in `#aggregate_column`

```ruby
def perform_calculation
  ....
  ....
  if operation == "count"
    column_name ||= select_for_count
    if column_name == :all
      ---> if distinct && !(has_limit_or_offset? && order_values.any?) <---
        column_name = primary_key
      end
    elsif column_name =~ /\s*DISTINCT[\s(]+/i
      distinct = nil
    end
  end
  ...
  ...
end
```

Perhaps this helps?

**Comment by kamipo** [comment]
Thanks for reporting the issue. I've fixed it and backported in 87ca68e76eec68ec61fc0d3d3bf315b4bfe45c42.
